src/Stratis.Bitcoin/Consensus/PHConsensusManagerBehavior.cs

```diff
@@ -0,0 +1,306 @@
+        {
+            ChainedHeader fork = this.chain.FindFork(locator);
+
+            lastHeader = null;
+
+            if (fork == null)
+            {
+                this.logger.LogTrace("(-)[INVALID_LOCATOR]:null");
+                return null;
+            }
+
+            var headers = new ProvenHeadersPayload();
+            var consensusFactory = new PosConsensusFactory();
+            foreach (ChainedHeader header in this.chain.EnumerateToTip(fork).Skip(1))
+            {
+                var posBock = new PosBlock(header.Header);
+                ProvenBlockHeader provenBlockHeader = consensusFactory.CreateProvenBlockHeader(posBock);
+                lastHeader = provenBlockHeader;
+                headers.Headers.Add(provenBlockHeader);
+
+                if ((header.HashBlock == hashStop) || (headers.Headers.Count == MaxItemsPerHeadersMessage))
+                    break;
+            }
+
+            return headers;
+        }
+
+        /// <summary>Validates the proven headers payload.</summary>
+        /// <param name="peer">The peer who sent the payload.</param>
+        /// <param name="provenHeadersPayload">Proven Headers payload to validate.</param>
+        /// <param name="validationError">The validation error that is set in case <c>false</c> is returned.</param>
+        /// <returns><c>true</c> if payload was valid, <c>false</c> otherwise.</returns>
+        private bool ValidateHeadersPayload(INetworkPeer peer, ProvenHeadersPayload provenHeadersPayload, out string validationError)
+        {
+            validationError = null;
+
+            if (provenHeadersPayload.Headers.Count > MaxItemsPerHeadersMessage)
+            {
+                this.logger.LogDebug("Proven Headers payload with {0} headers was received. Protocol violation. Banning the peer.", provenHeadersPayload.Headers.Count);
+
+                validationError = "Protocol violation.";
+
+                this.logger.LogTrace("(-)[TOO_MANY_HEADERS]:false");
+                return false;
+            }
+
+            // Check headers for consecutiveness.
+            for (int i = 1; i < provenHeadersPayload.Headers.Count; i++)
+            {
+                if (provenHeadersPayload.Headers[i].HashPrevBlock != provenHeadersPayload.Headers[i - 1].GetHash())
+                {
+                    this.logger.LogDebug("Peer '{0}' presented non-consecutiveness hashes at position {1} with prev hash '{2}' not matching hash '{3}'.",
+                        peer.RemoteSocketEndpoint, i, provenHeadersPayload.Headers[i].HashPrevBlock, provenHeadersPayload.Headers[i - 1].GetHash());
+
+                    validationError = "Peer presented nonconsecutive headers.";
+
+                    this.logger.LogTrace("(-)[NONCONSECUTIVE]:false");
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        protected override void AttachCore()
```

Comment:
If this is very similar to base you dont need to use it, as @noescape00 suggested just use override